### DISCUSSION TITLE
What is the meaning of the error: checksum missing, want 'SH2A56', got '' ?

### DISCUSSION THREAD

**Discussion by nowheresly** [minio/minio#17904] [Q&A] [answered]
Hi,

we're trying to upload an object in a bucket and we use the extra header "x-amz-checksum-sha256" to provide the sha-256 of the object we want to upload.

Now we get this error:

```
API: PutObjectPart
Time: 13:21:39 UTC 08/23/2023
DeploymentID: 5b8ae0c3-0ecc-41c2-9bd9-408a48d443a3
RequestID: 177E05CF97084DAC
RemoteHost: 10.152.8.23
UserAgent: MinIO (Linux; amd64) minio-java/8.5.5
Error: Invalid arguments provided for mybucket/dbc5092a6a4ec51c59e2d61d6092e8019dc9b3bcf18229c8f5b6e86d88385478/96b3317f-7db0-4b0b-a9eb-d6e73344f461: (checksum missing, want "SHA256", got "") (cmd.InvalidArgument)

Backtrace: 
0: internal/logger/logger.go:258:logger.LogIf()
1: cmd/api-errors.go:2308:cmd.toAPIErrorCode()
2: cmd/api-errors.go:2333:cmd.toAPIError()
3: cmd/object-multipart-handlers.go:818:cmd.objectAPIHandlers.PutObjectPartHandler()
4: net/http/server.go:2109:http.HandlerFunc.ServeHTTP()

```

Here is the code we use to upload the object on the client side:

```
                    minioClient.putObject(
                            PutObjectArgs.builder()
                                    .bucket("mybucket")
                                    .object("somefile.txt")
                                    .extraHeaders(Map.of("x-amz-checksum-sha256", checksum))
                                    .contentType("text/plain")
                                    .stream(is, 8192, ObjectWriteArgs.MIN_MULTIPART_SIZE)
                                    .build());

```

The bucket is an SSE-S3 encrypted bucket.

What is the meaning of this error ?
How can we fix the issue ?

Thanks

**Comment by klauspost** [comment, marked as answer]
You are triggering a multipart upload. Multiparts are a "checksum of checksum of part", not the combined stream SHA256.

From "Using part-level checksums for multipart uploads" here: https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html

> If you've enabled additional checksum values for your multipart object, Amazon S3 calculates the checksum for each individual part by using the specified checksum algorithm. The checksum for the completed object is calculated in the same way that Amazon S3 calculates the MD5 digest for the multipart upload. You can use this checksum to verify the integrity of the object.

Use a regular PutObject, which accepts a regular SHA256 or implement the S3 API.
